In [4]:
import pandas as pd
import matplotlib.pyplot as plt

df = pd.read_csv("../data/raw/Elimited_messy_sales_data.csv")
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,530331,22961,JUMBO BAG RED RETROSPOT,37,8/28/2025 17:08,2.08,17013.0,United Kingdom
1,530359,22457,CREAM CUPID HEARTS COAT HANGER,51,10/13/2025 11:48,2.75,13181.0,United Kingdom
2,530400,21212,PACK OF 72 RETROSPOT CAKE CASES,11,12/18/2025 16:59,0.55,17763.0,United Kingdom
3,530024,22720,POPCORN HOLDER,33,9/17/2025 3:55,0.85,17122.0,United Kingdom
4,C530303,22961,JUMBO BAG RED RETROSPOT,-2,8/29/2025 1:40,2.08,15169.0,United Kingdom


In [5]:
df.shape

(528, 8)

In [6]:
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='str')

In [7]:
df.dtypes

InvoiceNo          str
StockCode          str
Description        str
Quantity         int64
InvoiceDate        str
UnitPrice      float64
CustomerID     float64
Country            str
dtype: object

In [8]:
df.isnull().sum()

InvoiceNo       0
StockCode       0
Description     7
Quantity        0
InvoiceDate     0
UnitPrice       0
CustomerID     10
Country         0
dtype: int64

In [9]:
df.duplicated().sum()

np.int64(8)

In [10]:
df[["Quantity", "UnitPrice"]].describe()

,Quantity,UnitPrice
count,528.000000,528.000000
mean,70.212121,3.515265
std,583.141382,3.938864
min,-51.000000,0.000000
25%,14.000000,1.650000
50%,28.000000,2.080000
75%,42.000000,2.950000
max,9999.000000,18.000000


In [11]:
(df["Quantity"] < 0).sum()

np.int64(13)

In [12]:
(df["Quantity"] == 0).sum()

np.int64(4)

In [13]:
(df["UnitPrice"] == 0).sum()

np.int64(5)

In [14]:
df["InvoiceNo"].astype(str).str.startswith("C").sum()

np.int64(12)

In [15]:
df["Country"].unique()

<ArrowStringArray>
['United Kingdom',           'EIRE',        'GERMANY',       'PORTUGAL',
          'SPAIN',      'Australia',       'Portugal',           'Eire',
        'Belgium',          'Spain',        'BELGIUM',    'Switzerland',
         'France', 'UNITED KINGDOM',    'Netherlands',        'Germany',
         'FRANCE']
Length: 17, dtype: str

## Part 1: Data Understanding

- Dataset has 528 rows and 8 columns.
- Columns: InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country.
- InvoiceDate is stored as text, not a real date, this needs to be converted.
- CustomerID is stored as a decimal number because it has missing values.
- Description is missing in 7 rows.
- CustomerID is missing in 10 rows.
- There are 8 exact duplicate rows.
- 13 rows have negative Quantity, but only 12 InvoiceNo values start with "C" (the usual marker for cancelled orders). This means one negative quantity row does not follow the expected pattern and needs a closer look.
- 4 rows have Quantity equal to 0, which is not a valid sale.
- 5 rows have UnitPrice equal to 0, which is also not valid for a real transaction.
- The highest Quantity value is 9999, which is far above the typical range and looks like an error rather than a real order.
- Country names are inconsistent, some appear in normal case (e.g. "Germany") and others in all uppercase (e.g. "GERMANY"), which would cause the same country to be counted separately during analysis.